lib/rucio/daemons/common.py

```diff
@@ -23,22 +23,133 @@
 import time
 from collections.abc import Callable, Generator, Iterator, Sequence
 from typing import Any, Generic, Optional, TypeVar, Union
+from types import TracebackType, FrameType
+from abc import ABC, abstractmethod
 
-from rucio.common.logging import formatted_logger
+import rucio.db.sqla.util
+from rucio.common.logging import formatted_logger, setup_logging
 from rucio.common.utils import PriorityQueue
 from rucio.core import heartbeat as heartbeat_core
 from rucio.core.monitor import MetricManager
+from rucio.common.exception import DatabaseException
 
-T = TypeVar('T')
+T = TypeVar("T")
 METRICS = MetricManager(module=__name__)
 
 
+class Daemon(ABC):
+    """
+    Base daemon class
+    """
+
+    def __init__(
+        self,
+        once: bool = False,
+        total_workers: int = 1,
+        sleep_time: int = 60,
+        partition_wait_time: int = 1,
+        daemon_name: str = "undefined_daemon",
+    ) -> None:
+        """
+        :param once: Only execute one iteration.
+        :param total_workers: Number of total workers.
+        :param sleep_time: Sleep time between daemon iterations.
+        :param partition_wait_time: Time to wait for database partition rebalancing before starting the actual daemon loop.
+        :param daemon_name: Name of daemon that is constructed.
+        """
+        self.once = once
+        self.total_workers = total_workers
+        self.sleep_time = sleep_time
+        self.partition_wait_time = partition_wait_time
+        self.daemon_name = daemon_name
+        self.graceful_stop = threading.Event()
+        setup_logging(process_name=daemon_name)
+
+    @staticmethod
+    def _pre_run_checks() -> None:
+        """
+        Checks to run before daemon execution
+        """
+        if rucio.db.sqla.util.is_old_db():
+            raise DatabaseException("Database was not updated, daemon won't start")
+
+    @abstractmethod
+    def _run_once(
+        self, heartbeat_handler: "HeartbeatHandler", **_kwargs
+    ) -> tuple[bool, Any]:
+        """
+        Daemon-specific logic (to be defined in child classes) for a single iteration
+
+        :param heartbeat_handler: Handler to set and manage the heartbeat for this execution.
+
+        :returns: Tuple of (must_sleep, ret_value).
+                  must_sleep: set to True to signal to the calling context that it must sleep before next execution, False otherwise
+                  ret_value: Daemon-specific return value
+        """
+        pass
+
+    def _call_daemon(self, activities: Optional[list[str]] = None) -> None:
```

Comment:
I wonder if we should take the opportunity to rethink some of the chosen names. There’s ‘pre-run’, ‘run’, ‘run once’, ‘call daemon’. Do they feel intuitive? Had I not known, I would have guessed that `_call_daemon()` is the outer function, `_pre_run_checks()` is executed before `run()`, and that `run()` and `_run_once()` are alternatives.

Reply:
I am okay with changing the names if there are any suggestions, but I could not think of anything else at the moment. There are some things that I do not like, e.g. I think `once` and consequently `run_once` are not clear enough, but those names currently match some of the names from the daemon's user's perspective (e.g. the `--run-once` flag) so I didn't want to change them